### ISSUE TITLE
Fine-tune Any Models?

### ISSUE THREAD

**Issue by ZhiliWang** [patil-suraj/exploring-T5#1]
Hello Suraj, I have been really enjoying reading your code! I am wondering, for a beginner like me, is there any kind of tutorial I can learn to fine-tune almost any published model to do the certain task? e.g. fine-tuning T5 with SQuAD to answer open-book questions, which is using a context and a question to predict an answer. I think t5 has that task implemented but I am having a very difficult time doing so with non-Transformer models. 

Thank you!

**Comment by patil-suraj** [comment]
Hi @ZhiliWang  Sorry for the late response.
One disclaimer before I write my answer, I'm not the best person to ask this question but I'll try my best to answer.

First I would suggest you to get a basic idea of different problems in NLP and architectures. See what models are used for a particular task.

Once you choose the model depending on your task, take a look at the input and output representation of that model i.e how to encode the input and output. See which loss function to use for the task. Once you know what input the model takes, what it outputs and which loss function to use then you can easily setup a training loop.

Now if you want to use pre-trained models from Huggingface Transformers, then the best place to find this information is to look at docs. The docstring of  each model's forward method explains what the model expects as input and what it outputs.The docs has examples  of forward passes for every model. That should give you a good idea on how to train/fine-tune a particular model.  Once you do this few times you'll have a decent idea of how to fine-tune pre-trained models.

Also take a look at the examples in the Transformers repo. The [community notebooks](https://github.com/huggingface/transformers/tree/master/notebooks#community-notebooks) section is also very  helpful.  And if in doubt don't hesitate to open an issue and ask questions, the community there is really helpful.

Hope this helps you in some way :)
